### ISSUE TITLE
Destructure structs

### ISSUE THREAD

**Issue by LeonardJohard** [FluxML/Flux.jl#1380]
Collecting the gradient of this simple struct in a vector fails:

using Flux
struct test
  ann
  ann2
end
(m::test)(x) = m.ann(x) + m.ann2(x)
Flux.@functor test
newtest = test(Dense(1,1), Dense(1,1))
obj(test) = test([1.0])[1]
u1 = obj'(newtest)
Flux.destructure(u1)

This is true for any structure - it always collects one or more parameters too few unless there is just a single ann in the struct.

**Comment by CarloLucibello** [comment]
you should use github's code blocks for readability, and also display some output.

**Comment by ToucheSir** [comment]
`destructure` seems to collect the params just fine:
```julia
jjulia> ps, re = Flux.destructure(newtest)
(Float32[-1.0467398, 0.0, 0.9215399, 0.0], Flux.var"#34#36"{test}(test(Dense(1, 1), Dense(1, 1))))

julia> ps
4-element Array{Float32,1}:
 -1.0467398
  0.0
  0.9215399
  0.0

julia> re(ps)
test(Dense(1, 1), Dense(1, 1))
```
Is there a reason you're not using [`gradient`](https://fluxml.ai/Flux.jl/stable/models/basics/#Taking-Gradients-1) or `pullback` (or `train!`) to get gradients wrt. the parameters? `'` doesn't allow for parameterizing a loss function in terms of the input and model...

**Comment by LeonardJohard** [comment]
I am optimizing a bunch of complex structs of structs that did not work well for me with Params() + gradient().

I later got it working using Flatten anyhow. Zygote+Flatten naively seems like a promising way to extract parameters from more general AD, especially together with @functor from Flux to get convenient gpu functionality etc.

' is actually quite flexible. I am using something along the lines of:

loss(model, data, error) = error(model, data)
 loss(model) = loss(model, data, error)
flatten(model) .= flatten(model) .+ flatten(loss'(model))

where model is a nested struct containing arrays.

**Comment by ToucheSir** [comment]
Optimizing deeply nested structures is exactly what`@functor` was designed for! Either way, glad you managed to work it out. I'll second @CarloLucibello's recommendation to use triple backtick code blocks and include output for future issues (and issue comments) for clarity and a faster response.
